core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -578,61 +576,49 @@ public <T> T attr(AttributeKey<T> key) {
      */
     public <T> Endpoint withAttr(AttributeKey<T> key, @Nullable T value) {
         requireNonNull(key, "key");
-        if (value == null && attributes == null) {
-            return this;
+        if (attributes == null) {
+            if (value == null) {
+                return this;
+            }
+            return withAttrs(Attributes.of(key, value));
         }
-        if (attributes != null && attributes.attr(key) == value) {
+
+        if (attributes.attr(key) == value) {
             return this;
+        } else {
+            final AttributesBuilder attributesBuilder = attributes.toBuilder();
+            if (value == null) {
+                attributesBuilder.remove(key);
+            } else {
+                attributesBuilder.set(key, value);
+            }
+            return withAttrs(attributesBuilder.build());
         }
-        final DefaultAttributeMap newAttributes = new DefaultAttributeMap(null);
-        if (attributes != null) {
-            copyAttributes(newAttributes, attributes.attrs());
-        }
-        newAttributes.setAttr(key, value);
-        return withAttributes(newAttributes);
     }
 
     /**
-     * Returns a new host endpoint by adding the specified attributes.
-     *
-     * @return the new endpoint by adding the specified attributes. {@code this} if specified
-     *         attributes are empty.
+     * Returns a new {@link Endpoint} with the specified {@link Attributes}.
+     * Note that the {@link #attrs()} of this {@link Endpoint} is replaced with the specified
+     * {@link Attributes}.
      */
-    public Endpoint withAttrs(Iterable<? extends Entry<AttributeKey<?>, ?>> attributes) {
-        requireNonNull(attributes, "attributes");
-        final Iterator<? extends Entry<AttributeKey<?>, ?>> newAttrIterator = attributes.iterator();
-        if (!newAttrIterator.hasNext()) {
+    public Endpoint withAttrs(Attributes newAttributes) {
+        requireNonNull(newAttributes, "newAttributes");
+        if (newAttributes.isEmpty()) {
```

Comment:
So we are changing the behavior from append -> replace.

Should we still return as-is if `newAttributes.isEmpty()` instead of returning a new `Endpoint` with empty attributes?

Reply:
`Endpoint` is an immutable data structure. So I think we can return as-is.

Reply:
I see, I was wondering about this case:

```
Endpoint endpoint = origEndpoint.withAttributes(Attributes.of("a", "b"));
// regardless of what was in origEndpoint, now the attributes is overwritten to {a => b}

Endpoint endpoint = origEndpoint.withAttributes(Attributes.of());
// I expected endpoint now to have empty attributes, but it will have whatever attributes `origEndpoint` had
```

Let me know if I understood incorrectly 😅

Reply:
I see, I was wondering about this case:
```
Endpoint endpoint = origEndpoint.withAttributes(Attributes.of("a", "b"));
// regardless of what was in origEndpoint, now the attributes is overwritten to {a => b}

Endpoint endpoint = origEndpoint.withAttributes(Attributes.of());
// I expected endpoint now to have empty attributes, but it will have whatever attributes `origEndpoint` had
```
Let me know if I understood incorrectly 😅

Reply:
> // I expected endpoint now to have empty attributes, but it will have whatever attributes `origEndpoint` had

It must be a bug. 😅 It should return a new endpoint with an empty `Attributes`.